File: Source/Servers/Net/Nequeo.Net.Server/Nequeo.Net.Server/base/UdpServer.cs
/*  Company :       Nequeo Pty Ltd, http://www.nequeo.com.au/
 *  Copyright :     Copyright © Nequeo Pty Ltd 2014 http://www.nequeo.com.au/
 * 
 *  File :          
 *  Purpose :       
 * 
 */

#region Nequeo Pty Ltd License
/*
    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal in the Software without
    restriction, including without limitation the rights to use,
    copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the
    Software is furnished to do so, subject to the following
    conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
    OTHER DEALINGS IN THE SOFTWARE.
*/
#endregion

using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Text;
using System.Net;
using System.IO;
using System.Net.Security;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;

namespace Nequeo.Net
{
    /// <summary>
    /// Generic net UDP server.
    /// </summary>
    public class UdpServer : IDisposable
    {
        #region Constructors
        /// <summary>
        /// Generic net server.
        /// </summary>
        /// <param name="multiEndpointModels">The multi-endpoint model.</param>
        /// <param name="maxNumClients">The combined maximum number of clients for all servers with each endpoint.</param>
        public UdpServer(Net.Sockets.MultiEndpointModel[] multiEndpointModels, int maxNumClients = Int32.MaxValue)
        {
            _maxNumClients = maxNumClients;
            _multiEndpointModels = multiEndpointModels;
            Initialise();
        }
        #endregion

        #region Private Fields
        private Nequeo.Net.WebUdpServer _server = null;

        private int _maxNumClients = Int32.MaxValue;
        private Net.Sockets.MultiEndpointModel[] _multiEndpointModels = null;
        #endregion

        #region Public Properties
        /// <summary>
        /// Gets sets, the read buffer size.
        /// </summary>
        public int ReadBufferSize
        {
            get { return _server.ReadBufferSize; }
            set { _server.ReadBufferSize = value; }
        }

        /// <summary>
        /// Gets sets, the write buffer size.
        /// </summary>
        public int WriteBufferSize
        {
            get { return _server.WriteBufferSize; }
            set { _server.WriteBufferSize = value; }
        }

        /// <summary>
        /// Gets the socket server endpoints.
        /// </summary>
        public Nequeo.Net.WebUdpServer Servers
        {
            get { return _server; }
        }

        /// <summary>
        /// Gets, the is listening server indicator.
        /// </summary>
        public bool IsListening
        {
            get { return _server.IsListening; }
        }

        /// <summary>
        /// Gets or sets the timeout (seconds) for each client connection when in-active.
        /// Disconnects the client when this time out is triggered.
        /// </summary>
        public int Timeout
        {
            get { return _server.Timeout; }
            set { _server.Timeout = value; }
        }

        /// <summary>
        /// Gets or sets the length of time, in milliseconds, before the request times out; -1 wait indefinitely.
        /// </summary>
        public int RequestTimeout
        {
            get { return _server.RequestTimeout; }
            set { _server.RequestTimeout = value; }
        }

        /// <summary>
        /// Gets or sets the length of time, in milliseconds, before the response times out; -1 wait indefinitely.
        /// </summary>
        public int ResponseTimeout
        {
            get { return _server.ResponseTimeout; }
            set { _server.ResponseTimeout = value; }
        }

        /// <summary>
        /// Gets or sets the current server name.
        /// </summary>
        public string Name
        {
            get { return _server.Name; }
            set { _server.Name = value; }
        }

        /// <summary>
        /// Gets or sets the common service name.
        /// </summary>
        public string ServiceName
        {
            get { return _server.ServiceName; }
            set { _server.ServiceName = value; }
        }

        /// <summary>
        /// Gets or sets the maximum request buffer capacity when using buffered stream.
        /// </summary>
        public int RequestBufferCapacity
        {
            get { return _server.RequestBufferCapacity; }
            set { _server.RequestBufferCapacity = value; }
        }

        /// <summary>
        /// Gets or sets the maximum response buffer capacity when using buffered stream.
        /// </summary>
        public int ResponseBufferCapacity
        {
            get { return _server.ResponseBufferCapacity; }
            set { _server.ResponseBufferCapacity = value; }
        }
        #endregion

        #region Public Events
        /// <summary>
        /// The on web context event handler, triggered when a new connection is establised or data is present. Should be used when implementing a new connection.
        /// </summary>
        public event EventHandler<Nequeo.Net.WebContext> OnContext;

        #endregion

        #region Public Methods
        /// <summary>
        /// Start the server listener.
        /// </summary>
        public void Start()
        {
            _server.Start();
        }

        /// <summary>
        /// Stop the server listener.
        /// </summary>
        public void Stop()
        {
            _server.Stop();
        }
        #endregion

        #region Private Base server initialiser members
        /// <summary>
        /// Initialise.
        /// </summary>
        private void Initialise()
        {
            // Create the server.
            _server = new WebUdpServer(_multiEndpointModels, _maxNumClients);

            // Assign the on connect action handler.
            _server.Timeout = 60;
            _server.RequestTimeout = 30000;
            _server.ResponseTimeout = 30000;
            _server.ReadBufferSize = 32768;
            _server.WriteBufferSize = 32768;
            _server.ResponseBufferCapacity = 10000000;
            _server.RequestBufferCapacity = 10000000;
            _server.Name = "Generic Net UDP Server";
            _server.ServiceName = "GenericNetUdpServer";
            _server.OnWebContext += ManagerServer_OnWebContext;

            // Initialise the server.
            _server.Initialisation();
        }

        /// <summary>
        /// On web context receive handler.
        /// </summary>
        /// <param name="sender">The application sender.</param>
        /// <param name="context">The current web context.</param>
        private void ManagerServer_OnWebContext(object sender, Net.WebContext context)
        {
            // If the event has been assigned.
            if (OnContext != null)
                OnContext(this, context);
        }
        #endregion

        #region Dispose Object Methods
        /// <summary>
        /// Track whether Dispose has been called.
        /// </summary>
        private bool _disposed = false;

        /// <summary>
        /// Implement IDisposable.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            // This object will be cleaned up by the Dispose method.
            // Therefore, you should call GC.SuppressFinalize to
            // take this object off the finalization queue
            // and prevent finalization code for this object
            // from executing a second time.
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Dispose(bool disposing) executes in two distinct scenarios.  If disposing
        /// equals true, the method has been called directly or indirectly by a user's
        /// code. Managed and unmanaged resources can be disposed.  If disposing equals
        /// false, the method has been called by the runtime from inside the finalizer
        /// and you should not reference other objects. Only unmanaged resources can
        /// be disposed.
        /// </summary>
        protected virtual void Dispose(bool disposing)
        {
            // Check to see if Dispose has already been called.
            if (!this._disposed)
            {
                // Note disposing has been done.
                _disposed = true;

                // If disposing equals true, dispose all managed
                // and unmanaged resources.
                if (disposing)
                {
                    // Dispose managed resources.
                    if (_server != null)
                        _server.Dispose();
                }

                // Call the appropriate methods to clean up
                // unmanaged resources here.
                _server = null;
            }
        }

        /// <summary>
        /// Use C# destructor syntax for finalization code.
        /// This destructor will run only if the Dispose method
        /// does not get called.
        /// It gives your base class the opportunity to finalize.
        /// Do not provide destructors in types derived from this class.
        /// </summary>
        ~UdpServer()
        {
            // Do not re-create Dispose clean-up code here.
            // Calling Dispose(false) is optimal in terms of
            // readability and maintainability.
            Dispose(false);
        }
        #endregion
    }
}

File: Source/Servers/Net/NequeoNetServer/NequeoNetServer/UdpServer.cpp
/* Company :       Nequeo Pty Ltd, http://www.nequeo.com.au/
*  Copyright :     Copyright © Nequeo Pty Ltd 2016 http://www.nequeo.com.au/
*
*  File :          UdpServer.cpp
*  Purpose :       Udp Server class.
*
*/

/*
Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
*/

#pragma once

#include "stdafx.h"
#include "GlobalNetServer.h"

#include <boost/asio.hpp>

using boost::asio::ip::udp;

namespace Nequeo {
	namespace Net {
		namespace UDP
		{
			typedef std::function<void(const std::string&, const udp::endpoint&)> UdpReceiveHandler;

			/// <summary>
			///  UDP web server.
			/// </summary>
			class UdpServer
			{
			public:
				/// <summary>
				/// UDP web server.
				/// </summary>
				/// <param name="port">The UDP listening port.</param>
				/// <param name="protocol">The internet protocol.</param>
				UdpServer(unsigned int port, const boost::asio::ip::udp& protocol) :
					_disposed(false), _port(port), _initialised(false), _protocol(protocol)
				{
				}

				/// <summary>
				/// UDP web server.
				/// </summary>
				~UdpServer()
				{
					if (!_disposed)
					{
						_disposed = true;
						_initialised = false;
					}
				}

				/// <summary>
				/// Is initialise.
				/// </summary>
				/// <return>True if initialise; else false.</return>
				inline bool IsInitialise() const
				{
					return _initialised;
				}

				/// <summary>
				/// On received data.
				/// </summary>
				/// <param name="received">The received data handler.</param>
				void OnReceivedData(const UdpReceiveHandler& received)
				{
					_onReceivedData = received;
				}

				/// <summary>
				/// Initialise the servers.
				/// </summary>
				void Initialise()
				{
					// If not initialised.
					if (!_initialised)
					{
						// Create the new array of servers.
						_socket = std::make_shared<udp::socket>(_ioService, udp::endpoint(_protocol, _port));
						_initialised = true;

						// Start receiving data.
						StartReceive();
					}
				}

				///	<summary>
				///	Start the server.
				///	</summary>
				void Start()
				{
					// If initialised.
					if (_initialised)
					{
						try
						{
							// Main thread
							_ioService.run();
						}
						catch (...) {}
					}
				}

				///	<summary>
				///	Stop the server.
				///	</summary>
				void Stop()
				{
					// If initialised.
					if (_initialised)
					{
						try
						{
							// Close service.
							_ioService.stop();
						}
						catch (...) {}

						try
						{
							// Shutdown and close the client socket context.
							_socket->lowest_layer().shutdown(boost::asio::ip::tcp::socket::shutdown_both);
							_socket->lowest_layer().close();
							_socket->close();
						}
						catch (...) {}
					}
				}

				/// <summary>
				/// Send data to sender endpoint.
				/// </summary>
				/// <param name="data">The data to send.</param>
				/// <param name="sender">The sender endpoint.</param>
				void SendToSender(const std::string& data, const udp::endpoint& sender)
				{
					// Start sending data.
					_socket->async_send_to(boost::asio::buffer(data, data.size()), sender,
						[this](boost::system::error_code ec, std::size_t bytes_sent)
					{
						// Start receiving data.
						StartReceive();
					});
				}

			private:
				///	<summary>
				///	Start receiving data.
				///	</summary>
				void StartReceive()
				{
					udp::endpoint senderEndpoint;
					enum { max_length = 262144};
					char data[max_length];

					// Start receiving data.
					_socket->async_receive_from(boost::asio::buffer(data, max_length), senderEndpoint,
						[this, &senderEndpoint, &data](boost::system::error_code ec, std::size_t bytes_recvd)
					{
						// If data received.
						if (!ec && bytes_recvd > 0)
						{
							// Call the handler.
							_onReceivedData(std::string(data, bytes_recvd), senderEndpoint);
						}
						else
						{
							// Start receiving data.
							StartReceive();
						}
					});
				}

			private:
				bool _disposed;
				bool _initialised;

				unsigned int _port;
				boost::asio::ip::udp _protocol;
				std::shared_ptr<udp::socket> _socket;
				boost::asio::io_service _ioService;

				UdpReceiveHandler _onReceivedData;
			};

			/// <summary>
			///  UDP sender endpoint.
			/// </summary>
			class UdpSenderEndpoint
			{
			public:
				/// <summary>
				/// UDP sender endpoint.
				/// <param name="protocol">The internet protocol.</param>
				UdpSenderEndpoint() {}
				
				/// <summary>
				/// UDP sender endpoint.
				/// </summary>
				~UdpSenderEndpoint()
				{
					if (!_disposed)
					{
						_disposed = true;
					}
				}

				udp::endpoint senderEndpoint;
				std::string senderAddress;
				unsigned int senderPort;

			private:
				bool _disposed;
			};
		}
	}
}

